src/Compilers/Core/Portable/SourceGeneration/AdditionalSourcesCollection.cs

```diff
@@ -13,44 +14,108 @@ namespace Microsoft.CodeAnalysis
         }
 
-        internal AdditionalSourcesCollection(ImmutableArray<GeneratedSourceText> sources)
+        internal AdditionalSourcesCollection(ImmutableArray<GeneratedSourceText> existingSources)
             : this()
         {
-            foreach (var source in sources)
-            {
-                _sourcesAdded.Add(source.HintName, source.Text);
-            }
+            _sourcesAdded.AddRange(existingSources);
         }
 
         public void Add(string hintName, SourceText source)
         {
-            _sourcesAdded.Add(hintName, source);
+            if (string.IsNullOrWhiteSpace(hintName))
+            {
+                throw new ArgumentNullException(nameof(hintName));
+            }
+
+            // allow any identifier character or [.,-_ ()[]{}]
+            for (int i = 0; i < hintName.Length; i++)
+            {
+                char c = hintName[i];
+                if (!UnicodeCharacterUtilities.IsIdentifierPartCharacter(c)
+                    && c != '.'
+                    && c != ','
+                    && c != '-'
+                    && c != '_'
+                    && c != ' '
+                    && c != '('
+                    && c != ')'
+                    && c != '['
+                    && c != ']'
+                    && c != '{'
+                    && c != '}')
+                {
+                    throw new ArgumentException(string.Format(CodeAnalysisResources.HintNameInvalidChar, c, i), nameof(hintName));
+                }
+            }
+
+            hintName = AppendExtensionIfRequired(hintName);
+            lock (_collectionLock)
+            {
+                if (_hintNames.Contains(hintName))
+                {
+                    throw new ArgumentException(CodeAnalysisResources.HintNameUniquePerGenerator, nameof(hintName));
+                }
+
+                _hintNames.Add(hintName);
+                _sourcesAdded.Add(new GeneratedSourceText(hintName, source));
+            }
         }
 
         public void RemoveSource(string hintName)
         {
-            _sourcesAdded.Remove(hintName);
+            hintName = AppendExtensionIfRequired(hintName);
+            lock (_collectionLock)
+            {
+                if (_hintNames.Contains(hintName))
+                {
+                    _hintNames.Remove(hintName);
+                    for (int i = 0; i < _sourcesAdded.Count; i++)
+                    {
+                        // check the hashcodes, as thats the comparison we're using to put the names into _hintNames
+                        if (_sourcesAdded[i].HintName.GetHashCode() == hintName.GetHashCode())
+                        {
+                            _sourcesAdded.Remove(_sourcesAdded[i]);
+                            return;
+                        }
+                    }
+                }
+            }
         }
 
-        public bool Contains(string hintName) => _sourcesAdded.ContainsKey(hintName);
+        public bool Contains(string hintName)
```

Comment:
Don't like the name `Contains` because it's essentially a present tense word while the return value of this method could be invalid the moment it's calculated. Maybe `DidContain` or something of that sort.

Reply:
True. I renamed it `DidContain` it felt weird. All the other mutable collections have `Contains`, so going to leave as-is unless you feel _really_ strongly about it.

---
In reply to: [435563592](https://github.com/dotnet/roslyn/pull/44768#discussion_r435563592) [](ancestors = 435563592)

Reply:
In the case this collection is safe for concurrent read and write then I feel decently strongly about this. The most recent iteration though changed that to not be the case hence `Contains` is appropriate.